// @manekinekko/angular-web-bluetooth/esm2015/lib/logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function Logger() { }
if (false) {
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.log = function (args) { };
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.error = function (args) { };
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.warn = function (args) { };
}
export class ConsoleLoggerService {
    /**
     * @param {...?} args
     * @return {?}
     */
    log(...args) {
        console.log.apply(console, args);
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    error(...args) {
        console.error.apply(console, args);
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    warn(...args) {
        console.warn.apply(console, args);
    }
}
ConsoleLoggerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ ConsoleLoggerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ConsoleLoggerService_Factory() { return new ConsoleLoggerService(); }, token: ConsoleLoggerService, providedIn: "root" });
export class NoLoggerService {
    /**
     * @param {...?} args
     * @return {?}
     */
    log(...args) { }
    /**
     * @param {...?} args
     * @return {?}
     */
    error(...args) { }
    /**
     * @param {...?} args
     * @return {?}
     */
    warn(...args) { }
}
NoLoggerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ NoLoggerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NoLoggerService_Factory() { return new NoLoggerService(); }, token: NoLoggerService, providedIn: "root" });
