// @manekinekko/angular-web-bluetooth/esm5/lib/bluetooth.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bluetooth.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __awaiter, __generator, __read, __spread } from "tslib";
import { EventEmitter, Injectable } from '@angular/core';
import { from, fromEvent, throwError } from 'rxjs';
import { filter, map, mergeMap, takeUntil } from 'rxjs/operators';
import { ConsoleLoggerService } from './logger.service';
import { BrowserWebBluetooth } from './platform/browser';
import * as i0 from "@angular/core";
import * as i1 from "./platform/browser";
import * as i2 from "./logger.service";
var BluetoothCore = /** @class */ (function () {
    function BluetoothCore(webBle, console) {
        this.webBle = webBle;
        this.console = console;
        this.device$ = new EventEmitter();
        this.gatt$ = new EventEmitter();
        this.characteristicValueChanges$ = new EventEmitter();
        this.gattServer = null;
    }
    /**
     * @return {?}
     */
    BluetoothCore.prototype.getDevice$ = /**
     * @return {?}
     */
    function () {
        return this.device$;
    };
    /**
     * @return {?}
     */
    BluetoothCore.prototype.getGATT$ = /**
     * @return {?}
     */
    function () {
        return this.gatt$;
    };
    /**
     * @return {?}
     */
    BluetoothCore.prototype.streamValues$ = /**
     * @return {?}
     */
    function () {
        return this.characteristicValueChanges$.pipe(filter((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return value && value.byteLength > 0; })));
    };
    /**
     * Run the discovery process and read the value form the provided service and characteristic
     * @param options the ReadValueOptions
     */
    /**
     * Run the discovery process and read the value form the provided service and characteristic
     * @param {?} options the ReadValueOptions
     * @return {?}
     */
    BluetoothCore.prototype.value = /**
     * Run the discovery process and read the value form the provided service and characteristic
     * @param {?} options the ReadValueOptions
     * @return {?}
     */
    function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var device, gatt, primaryService, characteristic, value, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.console.log('[BLE::Info] Reading value with options %o', options);
                        if (typeof options.acceptAllDevices === 'undefined') {
                            options.acceptAllDevices = true;
                        }
                        if (typeof options.optionalServices === 'undefined') {
                            options.optionalServices = [options.service];
                        }
                        else {
                            options.optionalServices = __spread(options.optionalServices);
                        }
                        this.console.log('[BLE::Info] Reading value with options %o', options);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 7, , 8]);
                        return [4 /*yield*/, this.discover({
                                acceptAllDevices: options.acceptAllDevices,
                                optionalServices: options.optionalServices
                            })];
                    case 2:
                        device = (/** @type {?} */ (_a.sent()));
                        this.console.log('[BLE::Info] Device info %o', device);
                        return [4 /*yield*/, this.connectDevice(device)];
                    case 3:
                        gatt = _a.sent();
                        this.console.log('[BLE::Info] GATT info %o', gatt);
                        return [4 /*yield*/, this.getPrimaryService(gatt, options.service)];
                    case 4:
                        primaryService = (/** @type {?} */ (_a.sent()));
                        this.console.log('[BLE::Info] Primary Service info %o', primaryService);
                        return [4 /*yield*/, this.getCharacteristic(primaryService, options.characteristic)];
                    case 5:
                        characteristic = (/** @type {?} */ (_a.sent()));
                        this.console.log('[BLE::Info] Characteristic info %o', characteristic);
                        return [4 /*yield*/, characteristic.readValue()];
                    case 6:
                        value = _a.sent();
                        this.console.log('[BLE::Info] Value info %o', value);
                        return [2 /*return*/, value];
                    case 7:
                        error_1 = _a.sent();
                        throw new Error(error_1);
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @param {?} options
     * @return {?}
     */
    BluetoothCore.prototype.value$ = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return from(this.value(options));
    };
    /**
     * Run the discovery process.
     *
     * @param Options such as filters and optional services
     * @return  The GATT server for the chosen device
     */
    /**
     * Run the discovery process.
     *
     * @param {?=} options
     * @return {?} The GATT server for the chosen device
     */
    BluetoothCore.prototype.discover = /**
     * Run the discovery process.
     *
     * @param {?=} options
     * @return {?} The GATT server for the chosen device
     */
    function (options) {
        if (options === void 0) { options = (/** @type {?} */ ({})); }
        return __awaiter(this, void 0, void 0, function () {
            var device, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options.optionalServices = options.optionalServices || ['generic_access'];
                        this.console.log('[BLE::Info] Requesting devices with options %o', options);
                        device = null;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.webBle.requestDevice(options)];
                    case 2:
                        device = _a.sent();
                        device.addEventListener('gattserverdisconnected', this.onDeviceDisconnected.bind(this));
                        if (device) {
                            this.device$.emit(device);
                        }
                        else {
                            this.device$.error("[BLE::Error] Can not get the Bluetooth Remote GATT Server. Abort.");
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        error_2 = _a.sent();
                        this.console.error(error_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, device];
                }
            });
        });
    };
    /**
     * This handler will trigger when the client disconnets from the server.
     *
     * @param event The onDeviceDisconnected event
     */
    /**
     * This handler will trigger when the client disconnets from the server.
     *
     * @param {?} event The onDeviceDisconnected event
     * @return {?}
     */
    BluetoothCore.prototype.onDeviceDisconnected = /**
     * This handler will trigger when the client disconnets from the server.
     *
     * @param {?} event The onDeviceDisconnected event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var disconnectedDevice = (/** @type {?} */ (event.target));
        this.console.log('[BLE::Info] disconnected device %o', disconnectedDevice);
        this.device$.emit(null);
    };
    /**
     * Run the discovery process.
     *
     * @param Options such as filters and optional services
     * @return  Emites the value of the requested service read from the device
     */
    /**
     * Run the discovery process.
     *
     * @param {?=} options
     * @return {?} Emites the value of the requested service read from the device
     */
    BluetoothCore.prototype.discover$ = /**
     * Run the discovery process.
     *
     * @param {?=} options
     * @return {?} Emites the value of the requested service read from the device
     */
    function (options) {
        var _this = this;
        return from(this.discover(options)).pipe(mergeMap((/**
         * @param {?} device
         * @return {?}
         */
        function (device) { return _this.connectDevice$(device); })));
    };
    /**
     * Connect to current device.
     *
     * @return  Emites the gatt server instance of the requested device
     */
    /**
     * Connect to current device.
     *
     * @param {?} device
     * @return {?} Emites the gatt server instance of the requested device
     */
    BluetoothCore.prototype.connectDevice = /**
     * Connect to current device.
     *
     * @param {?} device
     * @return {?} Emites the gatt server instance of the requested device
     */
    function (device) {
        return __awaiter(this, void 0, void 0, function () {
            var gattServer, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!device) return [3 /*break*/, 5];
                        this.console.log('[BLE::Info] Connecting to Bluetooth Remote GATT Server of %o', device);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, device.gatt.connect()];
                    case 2:
                        gattServer = _a.sent();
                        this.gattServer = gattServer;
                        this.gatt$.emit(gattServer);
                        return [2 /*return*/, gattServer];
                    case 3:
                        error_3 = _a.sent();
                        // probably the user has canceled the discovery
                        Promise.reject("" + error_3.message);
                        this.gatt$.error("" + error_3.message);
                        return [3 /*break*/, 4];
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        this.console.error('[BLE::Error] Was not able to connect to Bluetooth Remote GATT Server');
                        this.gatt$.error(null);
                        _a.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Connect to current device.
     *
     * @return  Emites the gatt server instance of the requested device
     */
    /**
     * Connect to current device.
     *
     * @param {?} device
     * @return {?} Emites the gatt server instance of the requested device
     */
    BluetoothCore.prototype.connectDevice$ = /**
     * Connect to current device.
     *
     * @param {?} device
     * @return {?} Emites the gatt server instance of the requested device
     */
    function (device) {
        return from(this.connectDevice(device));
    };
    /**
     * Disconnect the current connected device
     */
    /**
     * Disconnect the current connected device
     * @return {?}
     */
    BluetoothCore.prototype.disconnectDevice = /**
     * Disconnect the current connected device
     * @return {?}
     */
    function () {
        if (!this.gattServer) {
            return;
        }
        this.console.log('[BLE::Info] Disconnecting from Bluetooth Device %o', this.gattServer);
        if (this.gattServer.connected) {
            this.gattServer.disconnect();
        }
        else {
            this.console.log('[BLE::Info] Bluetooth device is already disconnected');
        }
    };
    /**
     * Requests the primary service.
     *
     * @param gatt The BluetoothRemoteGATTServer sever
     * @param service The UUID of the primary service
     * @return The remote service (as a Promise)
     */
    /**
     * Requests the primary service.
     *
     * @param {?} gatt The BluetoothRemoteGATTServer sever
     * @param {?} service The UUID of the primary service
     * @return {?} The remote service (as a Promise)
     */
    BluetoothCore.prototype.getPrimaryService = /**
     * Requests the primary service.
     *
     * @param {?} gatt The BluetoothRemoteGATTServer sever
     * @param {?} service The UUID of the primary service
     * @return {?} The remote service (as a Promise)
     */
    function (gatt, service) {
        return __awaiter(this, void 0, void 0, function () {
            var remoteService, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 5]);
                        return [4 /*yield*/, gatt.getPrimaryService(service)];
                    case 1:
                        remoteService = _a.sent();
                        return [4 /*yield*/, Promise.resolve(remoteService)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        error_4 = _a.sent();
                        return [4 /*yield*/, Promise.reject(error_4.message + " (" + service + ")")];
                    case 4: return [2 /*return*/, _a.sent()];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Requests the primary service.
     *
     * @param gatt The BluetoothRemoteGATTServer sever
     * @param service The UUID of the primary service
     * @return The remote service (as an observable).
     */
    /**
     * Requests the primary service.
     *
     * @param {?} gatt The BluetoothRemoteGATTServer sever
     * @param {?} service The UUID of the primary service
     * @return {?} The remote service (as an observable).
     */
    BluetoothCore.prototype.getPrimaryService$ = /**
     * Requests the primary service.
     *
     * @param {?} gatt The BluetoothRemoteGATTServer sever
     * @param {?} service The UUID of the primary service
     * @return {?} The remote service (as an observable).
     */
    function (gatt, service) {
        this.console.log('[BLE::Info] Getting primary service "%s" (if available) of %o', service, gatt);
        if (gatt) {
            return from(this.getPrimaryService(gatt, service));
        }
        else {
            return throwError(new Error('[BLE::Error] Was not able to connect to the Bluetooth Remote GATT Server'));
        }
    };
    /**
     * Requests a characteristic from the primary service.
     *
     * @param primaryService The primary service.
     * @param characteristic The characteristic's UUID.
     * @returns The characteristic description (as a Promise).
     */
    /**
     * Requests a characteristic from the primary service.
     *
     * @param {?} primaryService The primary service.
     * @param {?} characteristic The characteristic's UUID.
     * @return {?} The characteristic description (as a Promise).
     */
    BluetoothCore.prototype.getCharacteristic = /**
     * Requests a characteristic from the primary service.
     *
     * @param {?} primaryService The primary service.
     * @param {?} characteristic The characteristic's UUID.
     * @return {?} The characteristic description (as a Promise).
     */
    function (primaryService, characteristic) {
        return __awaiter(this, void 0, void 0, function () {
            var char_1, rejectionError_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.console.log('[BLE::Info] Getting Characteristic "%s" of %o', characteristic, primaryService);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, primaryService.getCharacteristic(characteristic)];
                    case 2:
                        char_1 = _a.sent();
                        // listen for characteristic value changes
                        if (char_1.properties.notify) {
                            char_1.startNotifications().then((/**
                             * @param {?} _
                             * @return {?}
                             */
                            function (_) {
                                _this.console.log('[BLE::Info] Starting notifications of "%s"', characteristic);
                                char_1.addEventListener('characteristicvaluechanged', _this.onCharacteristicChanged.bind(_this));
                            }), (/**
                             * @param {?} error
                             * @return {?}
                             */
                            function (error) {
                                Promise.reject(error.message + " (" + characteristic + ")");
                            }));
                        }
                        else {
                            char_1.addEventListener('characteristicvaluechanged', this.onCharacteristicChanged.bind(this));
                        }
                        return [2 /*return*/, char_1];
                    case 3:
                        rejectionError_1 = _a.sent();
                        Promise.reject(rejectionError_1.message + " (" + characteristic + ")");
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Requests a characteristic from the primary service.
     *
     * @param primaryService The primary service.
     * @param characteristic The characteristic's UUID.
     * @returns The characteristic description (as a Observable).
     */
    /**
     * Requests a characteristic from the primary service.
     *
     * @param {?} primaryService The primary service.
     * @param {?} characteristic The characteristic's UUID.
     * @return {?} The characteristic description (as a Observable).
     */
    BluetoothCore.prototype.getCharacteristic$ = /**
     * Requests a characteristic from the primary service.
     *
     * @param {?} primaryService The primary service.
     * @param {?} characteristic The characteristic's UUID.
     * @return {?} The characteristic description (as a Observable).
     */
    function (primaryService, characteristic) {
        this.console.log('[BLE::Info] Getting Characteristic "%s" of %o', characteristic, primaryService);
        return from(this.getCharacteristic(primaryService, characteristic));
    };
    /**
     * Sets the characteristic's state.
     *
     * @param service The parent service of the characteristic.
     * @param characteristic The requested characteristic
     * @param state An ArrayBuffer containing the value of the characteristic.
     * @return The primary service (useful for chaining).
     */
    /**
     * Sets the characteristic's state.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic
     * @param {?} state An ArrayBuffer containing the value of the characteristic.
     * @return {?} The primary service (useful for chaining).
     */
    BluetoothCore.prototype.setCharacteristicState = /**
     * Sets the characteristic's state.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic
     * @param {?} state An ArrayBuffer containing the value of the characteristic.
     * @return {?} The primary service (useful for chaining).
     */
    function (service, characteristic, state) {
        var _this = this;
        /** @type {?} */
        var primaryService = this.getPrimaryService$(this.gattServer, service);
        primaryService
            // tslint:disable-next-line: variable-name
            .pipe(mergeMap((/**
         * @param {?} _primaryService
         * @return {?}
         */
        function (_primaryService) { return _this.getCharacteristic$(_primaryService, characteristic); })))
            // tslint:disable-next-line: no-shadowed-variable
            .subscribe((/**
         * @param {?} characteristic
         * @return {?}
         */
        function (characteristic) { return _this.writeValue$(characteristic, state); }));
        return primaryService;
    };
    /**
     * Enables the specified characteristic of a given service.
     *
     * @param service The parent service of the characteristic.
     * @param characteristic The requested characteristic
     * @return The primary service (useful for chaining).
     */
    /**
     * Enables the specified characteristic of a given service.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic
     * @param {?=} state
     * @return {?} The primary service (useful for chaining).
     */
    BluetoothCore.prototype.enableCharacteristic = /**
     * Enables the specified characteristic of a given service.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic
     * @param {?=} state
     * @return {?} The primary service (useful for chaining).
     */
    function (service, characteristic, state) {
        state = state || new Uint8Array([1]);
        return this.setCharacteristicState(service, characteristic, state);
    };
    /**
     * Disables the specified characteristic of a given service.
     *
     * @param service The parent service of the characteristic.
     * @param characteristic The requested characteristic.
     * @return The primary service (useful for chaining).
     */
    /**
     * Disables the specified characteristic of a given service.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic.
     * @param {?=} state
     * @return {?} The primary service (useful for chaining).
     */
    BluetoothCore.prototype.disbaleCharacteristic = /**
     * Disables the specified characteristic of a given service.
     *
     * @param {?} service The parent service of the characteristic.
     * @param {?} characteristic The requested characteristic.
     * @param {?=} state
     * @return {?} The primary service (useful for chaining).
     */
    function (service, characteristic, state) {
        state = state || new Uint8Array([0]);
        return this.setCharacteristicState(service, characteristic, state);
    };
    /**
     * Dispatches new values emitted by a characteristic.
     *
     * @param event the distpatched event.
     */
    /**
     * Dispatches new values emitted by a characteristic.
     *
     * @param {?} event the distpatched event.
     * @return {?}
     */
    BluetoothCore.prototype.onCharacteristicChanged = /**
     * Dispatches new values emitted by a characteristic.
     *
     * @param {?} event the distpatched event.
     * @return {?}
     */
    function (event) {
        this.console.log('[BLE::Info] Dispatching new characteristic value %o', event);
        /** @type {?} */
        var value = ((/** @type {?} */ (event.target))).value;
        this.characteristicValueChanges$.emit(value);
    };
    /**
     * Reads a value from the characteristics, as a DataView.
     *
     * @param characteristic The requested characteristic.
     * @return the DataView value (as an Observable).
     */
    /**
     * Reads a value from the characteristics, as a DataView.
     *
     * @param {?} characteristic The requested characteristic.
     * @return {?} the DataView value (as an Observable).
     */
    BluetoothCore.prototype.readValue$ = /**
     * Reads a value from the characteristics, as a DataView.
     *
     * @param {?} characteristic The requested characteristic.
     * @return {?} the DataView value (as an Observable).
     */
    function (characteristic) {
        this.console.log('[BLE::Info] Reading Characteristic %o', characteristic);
        return from(characteristic
            .readValue()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return Promise.resolve(data); }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return Promise.reject("" + error.message); })));
    };
    /**
     * Writes a value into the specified characteristic.
     *
     * @param characteristic The requested characteristic.
     * @param value The value to be written (as an ArrayBuffer or Uint8Array).
     * @return an void Observable.
     */
    /**
     * Writes a value into the specified characteristic.
     *
     * @param {?} characteristic The requested characteristic.
     * @param {?} value The value to be written (as an ArrayBuffer or Uint8Array).
     * @return {?} an void Observable.
     */
    BluetoothCore.prototype.writeValue$ = /**
     * Writes a value into the specified characteristic.
     *
     * @param {?} characteristic The requested characteristic.
     * @param {?} value The value to be written (as an ArrayBuffer or Uint8Array).
     * @return {?} an void Observable.
     */
    function (characteristic, value) {
        this.console.log('[BLE::Info] Writing Characteristic %o', characteristic);
        return from(characteristic.writeValue(value).then((/**
         * @param {?} _
         * @return {?}
         */
        function (_) { return Promise.resolve(); }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return Promise.reject("" + error.message); })));
    };
    /**
     * A stream of DataView values emitted by the specified characteristic.
     *
     * @param characteristic The characteristic which value you want to observe
     * @return The stream of DataView values.
     */
    /**
     * A stream of DataView values emitted by the specified characteristic.
     *
     * @param {?} characteristic The characteristic which value you want to observe
     * @return {?} The stream of DataView values.
     */
    BluetoothCore.prototype.observeValue$ = /**
     * A stream of DataView values emitted by the specified characteristic.
     *
     * @param {?} characteristic The characteristic which value you want to observe
     * @return {?} The stream of DataView values.
     */
    function (characteristic) {
        characteristic.startNotifications();
        /** @type {?} */
        var disconnected = fromEvent(characteristic.service.device, 'gattserverdisconnected');
        return fromEvent(characteristic, 'characteristicvaluechanged')
            .pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return (/** @type {?} */ (((/** @type {?} */ (event.target))).value)); })), takeUntil(disconnected));
    };
    /**
     * A utility method to convert LE to an unsigned 16-bit integer values.
     *
     * @param data The DataView binary data.
     * @param byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return An unsigned 16-bit integer number.
     */
    /**
     * A utility method to convert LE to an unsigned 16-bit integer values.
     *
     * @param {?} data The DataView binary data.
     * @param {?} byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return {?} An unsigned 16-bit integer number.
     */
    BluetoothCore.prototype.littleEndianToUint16 = /**
     * A utility method to convert LE to an unsigned 16-bit integer values.
     *
     * @param {?} data The DataView binary data.
     * @param {?} byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return {?} An unsigned 16-bit integer number.
     */
    function (data, byteOffset) {
        // tslint:disable-next-line:no-bitwise
        return (this.littleEndianToUint8(data, byteOffset + 1) << 8) + this.littleEndianToUint8(data, byteOffset);
    };
    /**
     * A utility method to convert LE to an unsigned 8-bit integer values.
     *
     * @param data The DataView binary data.
     * @param byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return An unsigned 8-bit integer number.
     */
    /**
     * A utility method to convert LE to an unsigned 8-bit integer values.
     *
     * @param {?} data The DataView binary data.
     * @param {?} byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return {?} An unsigned 8-bit integer number.
     */
    BluetoothCore.prototype.littleEndianToUint8 = /**
     * A utility method to convert LE to an unsigned 8-bit integer values.
     *
     * @param {?} data The DataView binary data.
     * @param {?} byteOffset The offset, in byte, from the start of the view where to read the data.
     * @return {?} An unsigned 8-bit integer number.
     */
    function (data, byteOffset) {
        return data.getUint8(byteOffset);
    };
    /**
     * Sends random data (for testing purposes only).
     *
     * @return Random unsigned 8-bit integer values.
     */
    /**
     * Sends random data (for testing purposes only).
     *
     * @param {?=} fakeValue
     * @return {?} Random unsigned 8-bit integer values.
     */
    BluetoothCore.prototype.fakeNext = /**
     * Sends random data (for testing purposes only).
     *
     * @param {?=} fakeValue
     * @return {?} Random unsigned 8-bit integer values.
     */
    function (fakeValue) {
        if (fakeValue === undefined) {
            fakeValue = (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var dv = new DataView(new ArrayBuffer(8));
                // tslint:disable-next-line:no-bitwise
                dv.setUint8(0, (Math.random() * 110) | 0);
                return dv;
            });
        }
        this.characteristicValueChanges$.emit(fakeValue());
    };
    BluetoothCore.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    BluetoothCore.ctorParameters = function () { return [
        { type: BrowserWebBluetooth },
        { type: ConsoleLoggerService }
    ]; };
    /** @nocollapse */ BluetoothCore.ɵprov = i0.ɵɵdefineInjectable({ factory: function BluetoothCore_Factory() { return new BluetoothCore(i0.ɵɵinject(i1.BrowserWebBluetooth), i0.ɵɵinject(i2.ConsoleLoggerService)); }, token: BluetoothCore, providedIn: "root" });
    return BluetoothCore;
}());
export { BluetoothCore };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.device$;
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.gatt$;
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.characteristicValueChanges$;
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.gattServer;
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.webBle;
    /**
     * @type {?}
     * @private
     */
    BluetoothCore.prototype.console;
}
