// @manekinekko/angular-web-bluetooth/esm5/lib/logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function Logger() { }
if (false) {
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.log = function (args) { };
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.error = function (args) { };
    /**
     * @param {?} args
     * @return {?}
     */
    Logger.prototype.warn = function (args) { };
}
var ConsoleLoggerService = /** @class */ (function () {
    function ConsoleLoggerService() {
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    ConsoleLoggerService.prototype.log = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.log.apply(console, args);
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    ConsoleLoggerService.prototype.error = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.error.apply(console, args);
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    ConsoleLoggerService.prototype.warn = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.warn.apply(console, args);
    };
    ConsoleLoggerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ ConsoleLoggerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ConsoleLoggerService_Factory() { return new ConsoleLoggerService(); }, token: ConsoleLoggerService, providedIn: "root" });
    return ConsoleLoggerService;
}());
export { ConsoleLoggerService };
var NoLoggerService = /** @class */ (function () {
    function NoLoggerService() {
    }
    /**
     * @param {...?} args
     * @return {?}
     */
    NoLoggerService.prototype.log = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    NoLoggerService.prototype.error = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
    };
    /**
     * @param {...?} args
     * @return {?}
     */
    NoLoggerService.prototype.warn = /**
     * @param {...?} args
     * @return {?}
     */
    function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
    };
    NoLoggerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ NoLoggerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NoLoggerService_Factory() { return new NoLoggerService(); }, token: NoLoggerService, providedIn: "root" });
    return NoLoggerService;
}());
export { NoLoggerService };
